plutus-core/plutus-core/src/PlutusCore/Crypto/ModExp.hs

```diff
@@ -0,0 +1,29 @@
+{-# LANGUAGE CPP         #-}
+{-# LANGUAGE MagicHash   #-}
+{-# LANGUAGE UnboxedSums #-}
+module PlutusCore.Crypto.ModExp
+    ( modularExponentiation
+    ) where
+
+import PlutusCore.Builtin
+
+import GHC.Natural
+
+#ifdef VERSION_ghc_bignum
+-- a.ka. ghc>=9 , a.k.a ghc-bignum added. Another way is to do MIN_VERSION_GLASGOW_HASKELL
+import GHC.Num.Integer
+
+modularExponentiation :: Integer -> Integer -> Natural -> BuiltinResult Natural
+modularExponentiation b e m =
+    case integerPowMod# b e m of
+        (# n | #)  -> pure n
+        (# | () #) -> fail "modularExponentiation: failure"
```

Comment:
@effectfully is it possible for a builtin function to return `Maybe`, or only builtin types can be returned?

Reply:
Depends on what we want `Nothing` to mean. If we want it to fail evaluation, then we can do that. If we want `Nothing` to become some kind of UPLC value, then no, unless we add `Maybe` to the universe or treat `[]` as `Nothing` (and `[a]` as `Just x`) or teach builtins about SOPs or...

I'd be against making `Nothing` mean failure, not only because it'd be weird to treat a common Haskell value as a UPLC exception, but also because every failure must come with an error message and `Maybe` doesn't have that functionality. So yeah, wanna fail a builtin -- you have to use `BuiltinResult` as your monad.